airflow/api_fastapi/execution_api/routes/task_instances.py

```diff
@@ -359,3 +364,23 @@ def ti_put_rtif(
     _update_rtif(task_instance, put_rtif_payload, session)
 
     return {"message": "Rendered task instance fields successfully set"}
+
+
+def _is_eligible_to_retry(task_instance: TI, task_retries: int | None):
```

Comment:
Two things;

1) This should probably be a method on TI. It might already exist as a method actually
2) I don't grok what we are doing with the `task_retries` parameter -- why do we need to pass it at all? When is `task_instance.try_number <= task_instance.max_tries` not right/good enough?

Reply:
For 1., there isn't a method on TI for this, I agree there should be one. Right now we have this https://github.com/apache/airflow/blob/main/airflow/models/taskinstance.py#L1054-L1073

Reply:
Regarding 2., looking at the code, I think its just safe coding.

We do this:
```
def _is_eligible_to_retry(*, task_instance: TaskInstance):
    """
    Is task instance is eligible for retry.

    :param task_instance: the task instance

    :meta private:
    """
    if task_instance.state == TaskInstanceState.RESTARTING:
        # If a task is cleared when running, it goes into RESTARTING state and is always
        # eligible for retry
        return True
    if not getattr(task_instance, "task", None):
        # Couldn't load the task, don't know number of retries, guess:
        return task_instance.try_number <= task_instance.max_tries

    if TYPE_CHECKING:
        assert task_instance.task

    return task_instance.task.retries and task_instance.try_number <= task_instance.max_tries
```

So I think we check in the last line, if there are retries defined and `task_instance.try_number <= task_instance.max_tries` is True, then retry. The reason is probably because retry can be defined as NONE.

```
retries: int | None = DEFAULT_RETRIES,
```

Reply:
Why do we do this though? That's what I don't get? With the advent of DAG versioning I don't think we can really get in the state where we can't find a task for a TI. And if we do: lets just error, if we can't find it then we can't re-run it really anyway.

Reply:
It has been taken care of now, yeah the point about not caring about "ti.task" makes sense.